### PR TITLE
Membership Audit

### LINKED ISSUE #168: Artisan command to reset and refresh the db
to bring up the db we need to do the following

```
php artisan doctrine:migration:refresh
php artisan permission:defaults
php artisan db:seed
php artisan passport:install
```

one command that wrapped theses and `doctrine:migration:reset` would be good,
also reset does not correctly remove all the oauth tables due to the filter

### PR DESCRIPTION
**Pull request by dpslwk**
Huge PR. but adding a lot of boiler plate and base tables

Added the following entities:

- Pin 
  - Pins a generated one first membership payment and allow registering of RFID cards with Gatekeeper
- RfidTag
- Zone
  - Doors are gateways between two zones
- Door
  - Gatekeeper controlled doors
- AccessLog
  - Keeps a record of people opening gatekeeper controlled doors
- Bank
  - The hackspace bank accounts
- BankTransaction
  - Imported bank records, matched against Account payment references.
- MembershipStatusNotification
  - Issues to members who have not paid us in the last 1 month and 14 days

The audit its self is triggered by an `AuditRequest` event, artisan command `hms:members:audit` can do this manually but the console kernel has is scheduled to run weekdays a 23:55

`AuditRequest` event is serviced by the `MembershipAudit` Listener which is where the bulk of the logic lies. This reviews the latest `BankTransaction` for each `Account` and determines if a User account needs moving between one of the Membership roles.

The audit will trigger the follow events / listeners for a member that needs to move between states

- NewMembershipPaidFor
  - ApproveNewMembership
    - Moves a member from `MEMBER_PAYMENT` to `MEMBER_CURRENT` or `MEMBER_YOUNG` based on the members DOB, generates a new `Pin` and fires off the `MembershipComplete` email to the User
- MembershipPaymentWarning
  - WarnMembershipMayExpire
    - Creates a new `MembershipStatusNotification` and fires off  a `MembershipMayBeRevoked` email to the user
- NonPaymentOfMembership
  - RevokeMembership
    - Moves a member form `MEMBER_CURRENT` or `MEMBER_YOUNG` to `MEMBER_EX`, sets any `MembershipStatusNotification` to REVOKED and fires off  a `MembershipRevoked` email to the user
- ReinstatementOfMembershipPayment
  - ReinstateMembership
    - Moves a member from `MEMBER_EX` to `MEMBER_CURRENT` or `MEMBER_YOUNG` based on the members DOB and fires off the `MembershipReinstated` email to the User

If the audit has any issues when processing users it will send a `AuditIssues` notification to the software team role

The `AuditResult` notification is sent to the Trustees and membership team roles on completion

closes #168

### CODE REVIEW

File: app/Listeners/Membership/WarnMembershipMayExpire.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    public function handle(MembershipPaymentWarning $event)
+    {
+        // get a fresh copy of the user
+        $user = $this->userRepository->find($event->user->getId());
+
+        $membershipStatusNotification = $this->membershipStatusNotificationFactory->create($user, $user->getAccount());
+        $this->membershipStatusNotificationRepository->save($membershipStatusNotification);
+
+        // emial user
```
Comment (geeksareforlife) [inline app/Listeners/Membership/WarnMembershipMayExpire.php]: email

File: app/Notifications/Banking/AuditIssues.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * Get the Slack representation of the notification.
+     *
+     * @param  mixed  $notifiable
+     * @return \Illuminate\Notifications\Messages\SlackMessage
+     */
+    public function toSlack($notifiable)
+    {
+        return (new SlackMessage)
```
Comment (geeksareforlife) [inline app/Notifications/Banking/AuditIssues.php]: This isn't included?

File: app/HMS/Entities/GateKeeper/Pin.php [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    /**
+     * Sets the value of state.
+     *
+     * @param int $state the state
+     *
+     * @return self
+     */
+    public function setState($state)
+    {
+        $this->state = $state;
```
[the remark this answers is not in the document]
Comment (dpslwk) [reply app/HMS/Entities/GateKeeper/Pin.php]: not easily done,

[the remark this answers is not in the document]
Comment (dpslwk) [reply app/HMS/Entities/GateKeeper/Pin.php]: nvm

File: app/HMS/Factories/Banking/BankTransactionFactory.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        $_bankTransaction = new BankTransaction();
+        $_bankTransaction->setBank($bank);
+        $_bankTransaction->setTransactionDate($transactionDate);
+        $_bankTransaction->setDescription($description);
+        $_bankTransaction->setAmount($amount);
+
+        if (preg_match('/HSNTSB\S{10}(?= )/', $description, $matches) == 1) {
+            $account = $this->accountRepository->findOneByPaymentRef($matches[0]);
+            $_bankTransaction->setAccount($account);
```
[the remark this answers is not in the document]
Comment (dpslwk) [reply app/HMS/Factories/Banking/BankTransactionFactory.php]: will just set null if not found

File: app/Notifications/Banking/AuditIssues.php (hunk 2 above)
Comment (dpslwk) [reply app/Notifications/Banking/AuditIssues.php]: fixed

**Comment by geeksareforlife** [comment]
Tested, works fine

**Pull request merged**